test/test_autograd.py

```diff
@@ -889,6 +889,162 @@ def test_grad_materialize_grads(self):
         with self.assertRaisesRegex(ValueError, "Expected allow_unused to be True or not passed when"):
             torch.autograd.grad(y, x, allow_unused=False, materialize_grads=True)
 
+    def test_post_accumulate_grad_hook_on_non_leaf_is_no_op(self):
+        def hook(tensor):
+            tensor.sub_(1.)
+        leaf = torch.rand(3, requires_grad=True)
+        non_leaf = 2. * leaf
+        non_leaf.register_post_accumulate_grad_hook(hook)
+        sum = non_leaf.sum()
+        sum.backward(retain_graph=True)
+        # The grad does not accumulate for non-leaf Tensors, even if
+        # retain_grad, so the hook should never be called.
+        self.assertEqual(non_leaf, leaf * 2.)
+
+    def test_post_accumulate_grad_hook_multiple_hooks(self):
+        def hook1(tensor):
+            tensor.sub_(tensor.grad)
+
+        def hook2(tensor):
+            tensor.mul_(4.)
+        tensor = torch.rand(3, requires_grad=True)
+        tensor_ref = tensor.clone().detach()
+        tensor.register_post_accumulate_grad_hook(hook1)
+        tensor.register_post_accumulate_grad_hook(hook2)
+        sum = tensor.sum()
+        sum.backward()
+        # both hooks should be called, in order
+        self.assertEqual(4. * (tensor_ref - 1.), tensor)
```

Comment:
This is pretty clever, though the other way to do it is to append something to a list and check the things in that list later (no opinion on this)

Reply:
nah i already thought out the math